openthread/src/core/common/message.hpp:122:38: error: ‘nonnull’ argument ‘aMessage’ compared to NULL [-Werror=nonnull-compare]

**Describe the bug** A clear and concise description of what the bug is.

Build failure with gcc 12.2.0

**To Reproduce** Information to reproduce the behavior, including:

1. Git commit id: 30d65bca4896a38e0e8a9e0b3cb0b9d458454ef2
2. IEEE 802.15.4 hardware platform: N/A
3. Build steps: `./bootstrap && ./script/cmake-build posix -DOT_SPINEL_RESET_CONNECTION=ON -DOT_THREAD_VERSION=1.2 -DOT_DAEMON=ON`

4. Network topology: N/A

**Expected behavior** A clear and concise description of what you expected to happen.

Build completes without warnings or errors

**Console/log output** If applicable, add console/log output to help explain your problem.

```
In file included from /home/zyga/projects/openthread/src/core/crypto/aes_ccm.hpp:43,
                 from /home/zyga/projects/openthread/src/core/thread/mle.hpp:45,
                 from /home/zyga/projects/openthread/src/core/thread/mle.cpp:34:
/home/zyga/projects/openthread/src/core/thread/mle.cpp: In member function ‘void ot::Mle::Mle::SendDelayedResponse(TxMessage&, const DelayedResponseMetadata&)’:
/home/zyga/projects/openthread/src/core/common/message.hpp:122:38: error: ‘nonnull’ argument ‘aMessage’ compared to NULL [-Werror=nonnull-compare]
  122 |         if (((aError) != kErrorNone) && ((aMessage) != nullptr)) \
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/zyga/projects/openthread/src/core/thread/mle.cpp:1643:5: note: in expansion of macro ‘FreeMessageOnError’
 1643 |     FreeMessageOnError(&aMessage, error);
      |     ^~~~~~~~~~~~~~~~~~
```

**Additional context** Add any other context about the problem here.

```
zyga@lambert:~/projects/openthread$ gcc --version
gcc (Ubuntu 12.2.0-1ubuntu1) 12.2.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

zyga@lambert:~/projects/openthread$ cmake --version
cmake version 3.24.1

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -186,11 +186,11 @@ jobs:
 
   gcc:
     name: gcc-${{ matrix.gcc_ver }}
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-22.04
     strategy:
       fail-fast: false
       matrix:
-        gcc_ver: [7, 8, 9, 10, 11]
+        gcc_ver: [9, 10, 11, 12]
     env:
       CC: gcc-${{ matrix.gcc_ver }}
       CXX: g++-${{ matrix.gcc_ver }}

--- a/src/core/net/dnssd_server.cpp
+++ b/src/core/net/dnssd_server.cpp
@@ -219,10 +219,10 @@ void Server::SendResponse(Header                  aHeader,
 
     error = aSocket.SendTo(aMessage, aMessageInfo);
 
-    FreeMessageOnError(&aMessage, error);
-
     if (error != kErrorNone)
     {
+        // do not use `FreeMessageOnError()` to avoid null check on nonnull pointer
+        aMessage.Free();
         LogWarn("failed to send DNS-SD reply: %s", ErrorToString(error));
     }
     else

--- a/src/core/thread/mle.cpp
+++ b/src/core/thread/mle.cpp
@@ -1640,7 +1640,11 @@ void Mle::SendDelayedResponse(TxMessage &aMessage, const DelayedResponseMetadata
     }
 
 exit:
-    FreeMessageOnError(&aMessage, error);
+    if (error != kErrorNone)
+    {
+        // do not use `FreeMessageOnError()` to avoid null check on nonnull pointer
+        aMessage.Free();
+    }
 }
 
 void Mle::RemoveDelayedDataResponseMessage(void)

--- a/src/ncp/ncp_hdlc.hpp
+++ b/src/ncp/ncp_hdlc.hpp
@@ -128,9 +128,9 @@ private:
                                                       Spinel::Buffer *         aBuffer);
     otNcpHdlcSendCallback mSendCallback;
 
+    Hdlc::FrameBuffer<kHdlcTxBufferSize> mHdlcBuffer;
     Hdlc::Encoder                        mFrameEncoder;
     Hdlc::Decoder                        mFrameDecoder;
-    Hdlc::FrameBuffer<kHdlcTxBufferSize> mHdlcBuffer;
     HdlcTxState                          mState;
     uint8_t                              mByte;
     Hdlc::FrameBuffer<kRxBufferSize>     mRxBuffer;

